src/karmen_frontend/src/app.js

```diff
@@ -50,183 +61,328 @@ const ScrollToTop = () => {
-        this.setState({
-          initialized: true,
-        });
-      });
-    } else {
-      this.myRef.current.scrollTo(0, 0);
+  Sentry.withScope((scope) => {
+    scope.setExtras(error);
+    sentryEventId = Sentry.captureException(error);
+  });
+
+  return (
+    <div role="alert" className="content">
+      <div className="container text-center">
+        <h1 className="main-title text-center">Something didn't go well</h1>
+        <p>
+          We're sorry, but there has been an error in the application.
+          <br />
+          Our engineers were already notified and will fix it as soon as
+          possible.
+        </p>
+        <div className="cta-box">
+          <button className="btn" type="reset" onClick={resetErrorBoundary}>
+            Try again
+          </button>
+          {sentryEventId && (
+            <button
+              className="btn btn-plain"
+              onClick={() =>
+                Sentry.showReportDialog({ eventId: sentryEventId })
+              }
+            >
+              Submit error report
+            </button>
+          )}
+        </div>
+      </div>
+    </div>
+  );
+};
+
+/**
+ * This is a general handler for uncaught promise rejections. It will be attached to
+ * window object when <AppErrorGuard> renders, see below.
+ */
+const rejectionErrorHandler = (event) => {
+  const err = event.reason;
+
+  if (err instanceof HttpError) {
+    switch (err.response.status) {
+      case 400:
+        return toast(
+          <>
+            <div className="toast-main">Something didn't go well.</div>
+            <p>
+              There has been some trouble talking to the backend server. Sorry!
+            </p>
+          </>,
+          { autoClose: false, toastId: "HttpError400" }
+        );
+      case 401:
+        return toast(
+          <>
+            <div className="toast-main">
+              Your session has expired. Please log in again.
+            </div>
+            <p>This usually happens after long period of inactivity.</p>
+          </>,
+          { autoClose: 8000, toastId: "HttpError401" }
+        );
+      case 403:
+        return toast(
+          <>
+            <div className="toast-main">
+              You're not allowed to display this page.
+            </div>
+            <p>You lack the required permissions.</p>
+          </>,
+          { autoClose: false, toastId: "HttpError403" }
+        );
+      default:
+        return toast(
+          <>
+            <div className="toast-main">That's an error!</div>
+            <p>An unknown error occured on the server. Sorry!</p>
+          </>,
+          { autoClose: false, toastId: "HttpErrorOther" }
+        );
     }
   }
 
-  componentWillUnmount() {
-    const { timer } = this.state;
-    timer && clearTimeout(timer);
-  }
+  toast(
+    <>
+      <div className="toast-main">That's an error!</div>
+      <p>Yep, there's been some problem on our end. Sorry!</p>
```

Comment:
I'm not sure about showing the general error for every failed connection. If I e.g. lost WiFi signal and go offline, I get the offline message, as request to `/api` fails. But the periodical checks for printers status are still alive in background, and after some time, they cause this message to pop up as they also fail. Maybe do not show them if the big Offline banner is displayed? 

![Screenshot from 2020-05-27 16-15-56](https://user-images.githubusercontent.com/10603257/83031830-5f409400-a035-11ea-8497-4e0009172ec8.png)

Reply:
Yes, that definitely needs to be sorted out.

Reply:
Also, If I open an URL to organization I'm not a member of, I get redirected to 404 page (which is fine) and then this error also shows.

Reply:
I've tried this locally and it only reports error when BE is down indeed. No toast is fired in offline-only scenario. Are you sure it's shown in your case?

Reply:
This doesn't show for hearbeat failure (e.g. general offline state). It gets displayed when loading printer list fails though as you suggested. Probably, we will need to do it as you say, to monitor offline state and do not show such errors if we're offline.

Reply:
I put some more thought in it and I suggest following:

- Add heartbeat to the store so that individual components can reflect online/offline state consistently.
- We shouldn't fire off HTTP requests when app is offline (except for heartbeat check) in the first place. Once connection status is available everywhere, we should skip HTTP requests completely for offline state with a graceful `OfflineError` for example.

Reply:
This has been addressed in 2ec2c432d82aec39bc61b3701025d52facdf2cc8, e0992daf5b56cc9c18ae628e842e6a72403b13d4 and e4f936fafc2c20a4096047d2f4d1602de47bfd22. Please have a look @adamjezek98 .